watcher/src/watcher_db.rs

```diff
@@ -165,6 +165,40 @@ impl WatcherDB {
             .collect::<Result<Vec<_>, WatcherDBError>>()?)
     }
 
+    /// Get the earliest timestamp for a given block.
+    /// The earliest timestamp reflects the time closest to when the block passed consensus.
+    /// If no timestamp is present, return u64::MAX, and a status code.
+    ///
+    /// Note: If there are no Signatures (and therefore no timestamps) for the given
+    ///       block, the result from get_signatures will be Ok(vec![]).
+    ///       A consensus validator only writes a signature for a block in which it
+    ///       participated in consensus. Therefore, if the watcher is only watching
+    ///       a subset of nodes, and those nodes happened to not participate in this
+    ///       block, the timestamp result will be unavailable for this block. It is
+    ///       also possible to be in a temporary state where there are no signatures
+    ///       for the given block if the watcher sync is behind the ledger sync.
+    pub fn get_block_timestamp(
+        &self,
+        block_index: u64,
+    ) -> Result<(u64, TimestampResultCode), WatcherDBError> {
+        if block_index == 0 {
+            return Ok((u64::MAX, TimestampResultCode::BlockIndexOutOfBounds));
+        }
+        let sigs = self.get_block_signatures(block_index)?;
+        match sigs.iter().map(|s| s.block_signature.signed_at()).min() {
+            Some(earliest) => Ok((earliest, TimestampResultCode::TimestampFound)),
+            None => {
+                // Check whether we are synced for all watched URLs
+                let highest_synced = self.highest_synced_block()?;
```

Comment:
I think this still suffers from the issue mentioned awhile ago where if the watcher is restarted with a different set of URLs that results in a URL being removed, `highest_synced_block` will get stuck at that point in time.

Reply:
Yep - that remains the expected behavior - I have a ticket for addressing this once the timestamp APIs are complete: https://mobilecoin.atlassian.net/browse/MC-1641

Reply:
Resolved thanks to @eranrund in #312